dsgrid/config/project_config.py

```diff
@@ -730,43 +732,142 @@ def _iter_datasets_by_status(self, status: DatasetRegistryStatus) -> InputDatase
-            )
+            ).cache()
+
+        # Building the table in order of increasing unique counts was the fastest in one test.
+        # That may not always be the case. If we don't pick an order, it could be different
+        # each time, which isn't good. So, go with this.
+        keys = sorted(other_dims.keys(), key=lambda x: other_dims[x].count())
+        for dim in keys:
+            other = other_dims[dim]
+            logger.info("Start processing dimension=%s table_count=%s", dim, table_count)
             if table is None:
                 table = other
             else:
                 table = table.crossJoin(other)
-                assert table.count() == other.count() * table_count
+                count = table.count()
+                assert count == other.count() * table_count
+                table_count = count
+            other.unpersist()
+
+        logger.info("Dimension association table_count=%s", table_count)
+        return table
+
+    def _replace_supplemental_records_with_base_records(self, table):
+        supp_dim_query_names = self.get_supplemental_dimension_to_query_name_mapping()
+        # TODO DT: assert that query names aren't data_sources
+        exclude = (DimensionType.DATA_SOURCE, DimensionType.TIME)
+        for dim_type in (x for x in DimensionType if x not in exclude):
+            unique_ids = get_unique_values(table, dim_type.value)
+            if unique_ids:
+                base_records = self.get_base_dimension(dim_type).get_unique_ids()
+                supplemental_ids = unique_ids.difference(base_records)
+                for supplemental_id in supplemental_ids:
+                    table = self._replace_supplemental_record_with_base_records(
+                        table,
+                        supp_dim_query_names,
+                        dim_type,
+                        supplemental_id,
+                    )
+        return table
+
+    def _replace_supplemental_record_with_base_records(
+        self, table, supp_dim_query_names, dim_type, supplemental_id
+    ):
+        found = False
+        for query_name in supp_dim_query_names[dim_type]:
+            mapping_records = (
+                self.get_base_to_supplemental_mapping_records(query_name)
+                .select("from_id", "to_id")
+                .filter(f"to_id == '{supplemental_id}'")
+                .cache()
+            )
+            if not mapping_records.rdd.isEmpty():
+                if found:
+                    raise DSGInvalidDimensionAssociation(
+                        f"dimension association for {dim_type} has supplemental record ID "
+                        "{supplemental_id} defined in at least two supplemental dimensions"
+                    )
+                found = True
+                new_records = (
+                    table.join(
+                        mapping_records,
+                        on=table[dim_type.value] == mapping_records.to_id,
```

Comment:
I think this is a one-to-many join, with one value `table[dim_type.value]` mappable to many values in `mapping_records.to_id`. Did the inner join produce  join as expected? And do you need to prefilter `mapping_records` before the inner join here?

Reply:
What happens if you do table.left_join(mapping_records)? Could you get the new table directly without doing the union below?